### PR TITLE
Basic functionality for the notification system

### PR DESCRIPTION
**Pull request by countgizmo**
So far the Notifikator can do:
- trigger messages by event
- dismiss messages by clicking the close icon
- stack multiple messages
- includes a playground for demo/testing/showing-off

### CODE REVIEW

File: src/cljs/notifikator/events.cljs [hunk 1]
```diff
@@ -1,9 +1,43 @@
...
+  (let [id (next-msg-id db)
+        msg {:id id
+             :title title
+             :description description
+             :class flavor}]
+    (update db :messages conj msg)))
+
+(re-frame/reg-event-db
+  ::spawn-message
+  (fn [db effect flavor]
```
Comment (volrath) [inline src/cljs/notifikator/events.cljs]: `reg-event-db` handler functions [only receive 2 arguments](https://github.com/Day8/re-frame/blob/master/src/re_frame/core.cljc#L75): `db` and `event-vector`, or in the more generic `reg-event-fx`, you receive `coeffects` and `event-vector`.

File: src/cljs/notifikator/events.cljs [hunk 2]
```diff
@@ -1,9 +1,43 @@
...
+
+(re-frame/reg-event-db
+  ::spawn-error-message
+  (fn [db effect]
+    (handle-message db effect "error message")))
+
+(re-frame/reg-event-db
+  ::close-message
+  (fn [db [_ id]]
+    (update-in db [:messages] remove-by-id id)))
```
Comment (volrath) [inline src/cljs/notifikator/events.cljs]: maybe use plain `update`?

File: src/cljs/notifikator/util.cljs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+(defn next-msg-id
+  "Return 0 if there are no messages.
+   If not empty, find the max ID and return its incremente value."
+  [{:keys [messages]}]
+  (if (or (empty? messages)
+          (nil? messages))
+    0
+    (-> (apply max-key :id messages)
+        :id
+        inc)))
```
Comment (volrath) [inline src/cljs/notifikator/util.cljs:18]: minor: you could also consider using a [random uuid](https://github.com/lbradstreet/cljs-uuid-utils)

File: src/cljs/notifikator/views.cljs [hunk 4]
```diff
@@ -1,8 +1,79 @@
...
+  ^{:key (str "msg-" id)}
+  [:div {:class class}
+   [:div {:class "title"} title (close-button id)]
+   [:div {:class "description"} description]])
+
+(defn messages-container
+  "If there are messages in the app-state create a message component
+   for each of them. Otherwise return nil. Which is fine for Hiccup."
+  []
+  (when-let [messages (re-frame/subscribe [::subs/messages])]
```
Comment (volrath) [inline src/cljs/notifikator/views.cljs]: This `when-let`'s condition is always going to be true, cause `re-frame/subscribe` returns (something-like-) a reagent's `atom`.

Comment (volrath) [reply src/cljs/notifikator/views.cljs]: Also, even though this works ok, the reagent's "form 1, 2, 3 component" rules also apply here.

File: src/cljs/notifikator/views.cljs [hunk 5]
```diff
@@ -1,8 +1,79 @@
...
+         [:textarea {:value @description
+                     :on-change #(reset! description (-> % .-target .-value))}]]]
+       [:div.buttons
+        [generate-info-message @title @description]
+        [generate-warning-message @title @description]
+        [generate-error-message @title @description]]])))
+
+(defn test-view
+  []
+  ^{:key "wold-is-a-simulation"}
```
Comment (volrath) [inline src/cljs/notifikator/views.cljs]: minor: from a React perspective, this `:key` might not be needed, I think.

**Review by volrath — CHANGES_REQUESTED** [review]
Good job!

File: src/cljs/notifikator/views.cljs (hunk 5 above)
Comment (countgizmo) [reply src/cljs/notifikator/views.cljs]: Thought so as well but then got a warning in the console.  After adding this key the warning was gone. I will try again, cause now I also have key for each message so maybe it was actually bubbling up from there.

File: src/cljs/notifikator/views.cljs (hunk 4 above)
Comment (countgizmo) [reply src/cljs/notifikator/views.cljs]: Thanks! Did not know that - still learning re-frame :)

File: src/cljs/notifikator/util.cljs (hunk 3 above)
Comment (countgizmo) [reply src/cljs/notifikator/util.cljs:18]: Actually a good idea. Especially considering I can remove messages in any order. Can also drop the next-id calculation logic. But I'll keep my simple IDs for now. Two reasons:
1. Don't want an extra dependency in this small POC task.
2. Smaller and easier to read in HTML/React dev tools.

But yeah, for actual prod I would probably go with random UUIDs to exclude the complexity of tracking the next ID and thus potential bugs.

File: src/cljs/notifikator/events.cljs (hunk 2 above)
Comment (countgizmo) [reply src/cljs/notifikator/events.cljs]: Oh! Yeah, you're right: it's not nested. I think I was trying to do something tricky with update-in but then failed and just created a remove-by-id function then didn't notice that I don't need to go nested anymore. Will change that.

File: src/cljs/notifikator/events.cljs (hunk 1 above)
Comment (countgizmo) [reply src/cljs/notifikator/events.cljs]: This is a great example of what can happen by the end of my workday: brain starts skipping cycles :) Thanks for pointing this out! :)

File: src/cljs/notifikator/views.cljs (hunk 5 above)
Comment (countgizmo) [reply src/cljs/notifikator/views.cljs]: Yeah, it seems the error was bubbling up from messages. Now when messages have keys this is not needed at all. Cleaned-up.

**Comment by countgizmo** [comment]
@volrath valuable stuff, thanks! I've pushed a commit that addresses all of them (except the UUID). Please let me know if you're happy with the changes or I've missed something else.

**Pull request merged**